tutorial/python_agents/base_agent.py

```diff
@@ -0,0 +1,485 @@
+            elif signal[3] == "avg":
+                if delta_time == 0:
+                    # Explicitly doing this allows the final value to be 0 even if
+                    # processed_sigs[idx] was 0 (i.e. first sample).
+                    processed_sigs[idx] = 0
+                else:
+                    processed_sigs[idx] *= delta_time
+
+        self._signal_accumulator = [aa + bb
+                                    for (aa, bb)
+                                    in zip(self._signal_accumulator, processed_sigs)]
+
+        ret = self.agent_update(delta_sigs, delta_time, signals[:-1], signals[-1])
+        geopmpy.reporter.update()
+
+        if len(ret) == 2:
+            controls, trace_fields = ret
+            debug_fields = []
+        elif len(ret) == 3:
+            controls, trace_fields, debug_fields = ret
+
+        if self._trace_file is not None and trace_fields is not None:
+            self._trace_file.write(
+                CSV_SEP_CHAR.join(
+                    [str(signal) for signal in trace_fields] +
+                    [str(control) for control in controls] +
+                    [str(field) for field in debug_fields]) +
+                "\n")
+
+        self._last_signals = list(signals[:-1])
+        self._last_time = signals[-1]
+        self._loop_idx += 1
+
+        return controls
+
+    def loop_idx(self):
+        """
+        Retuns
+
+        The count of how many times update method was called on the agent by the controller.
+        """
+        return self._loop_idx
+
+    def agent_update(self, delta_signals, delta_time, signals, time):
+        """This method is called once per update after the diff (delta) values are computed for
+        signals. The default implementation is a noop but an extending class can make decisions
+        on what controls to apply and what to print in the trace line. If this method returns
+        control signal values, get_controls method need to be overridden appropriately.
+
+        Args
+
+        delta_signals (list of float): Diffed version of the signals that are marked as incr,
+        actual signal value for signals marked as avg. In the same order as the signals
+        in signal_list passed to the class constructor.
+
+        delta_time (float): Diffed version of the time signal, in seconds.
+
+        signals (list of float): Signal values as they are passed to the update method of BaseAgent.
+        In the same order as the signals in signal_list passed to the class constructor.
+
+        time (float): TIME signal value as it is passed to the update method of BaseAgent.
+
+        Returns
+
+        A tuple of controls and trace values.
+
+        controls (list of float): Value per signal in get_controls method.
+
+        trace_values (list of float): Value per signal in get_signals method prepended by the value
+        for the TIME column.
+        """
+        return [], [delta_time] + delta_signals
```

Comment:
```suggestion
        return [], [time] + delta_signals
```

Using [time] results in a trace closer to the C++ tracer output.